src/H5Tnative.c

```diff
@@ -715,7 +717,14 @@ H5T__get_native_float(size_t size, H5T_direction_t direction, size_t *struct_ali
     assert(size > 0);
 
     if (direction == H5T_DIR_DEFAULT || direction == H5T_DIR_ASCEND) {
-        if (size <= sizeof(float)) {
+#ifdef H5_HAVE__FLOAT16
+        if (size <= sizeof(H5__Float16)) {
```

Comment:
It seems dangerous for the code here to match smaller datatypes with larger datatypes, since, before this change, the test code we currently have in ntypes.c was matching `_Float16` to `float` and causing an `H5Dread` call using the type returned from `H5Tget_native_type` to overrun the buffer. Seems to me that we should only match a type if the sizes match. That said, I'm just keeping in line with how the code behaved previously here.

Reply:
I think that's a bug in the test code. `H5Tget_native_type` doesn't guarantee the size will match, it kind of fuzzily selects the closest native match to the datatype passed in. The app code should inspect the datatype returned to determine how to allocate the buffer and interpret the data. (If a real app knew the size a priori, it probably wouldn't need to call `H5Tget_native_type` anyways).